EventHubs, Sample13, Client Secret Authentication

albertodenatale opened the pull request:
Hello @jsquire 
Hope you are well,

As we discussed I am advancing a pull request with sample 13 and some tweaks to the infrastructure to accommodate it.

Commit:
[Event Hubs Client] Samples 12, illustration of how to use Event Hubs with Azure.Identity and ClientSecretCredential in order to send and consume events

azuresdkci commented:
Can one of the admins verify this patch?

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample13_AuthenticateWithClientSecretCredential.cs:
```diff
@@ -0,0 +1,71 @@
+        ///   A short description of the sample.
+        /// </summary>
+        ///
+        public string Description { get; } = "An example of authentication using ClientSecretCredential.";
```
A bit more context may be helpful.  Perhaps something like:  `An example of creating an Event Hub client using a Azure Active Directory application with a client secret for authorization.` ?

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample13_AuthenticateWithClientSecretCredential.cs:
```diff
@@ -0,0 +1,71 @@
+                                   string clientId,
+                                   string secret)
+        {
+            // Service principal authentication is a mean for applications to authenticate
```
nit:  (typo) `means`

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample13_AuthenticateWithClientSecretCredential.cs:
```diff
@@ -0,0 +1,71 @@
+            // Service principal authentication is a mean for applications to authenticate
+            // against Azure Active Directory (AAD) and consume Azure services. This is advantageous compared
+            // to signing in using fully privileged users as it allows to enforce role-based authorization
+            // from the portal.
```
It also allows for rolling of credentials without tearing down clients that may be running - something that is difficult if not impossible for connection strings.

albertodenatale replied:
I am not gonna lie to you, I think I lost the plot here. I saw this [issue](https://github.com/Azure/azure-sdk-for-js/issues/1732) does this mean that if I am changing an SAS key in a connection string AMQP connections with all the clients will be disrupted while if I change the secret connected to a service principal all the clients will be able to communicate until their token won't expire?

jsquire replied:
There's a couple of things:

- Rolling credentials for a token-based user is something that was designed to be supported without disrupting an application.   As a client library, we want to encourage folks to use the Identity-based paths for production applications in the name of best practices for security and also for stability of the application.

- You're correct in that rolling credentials for the shared key in the portal will cause existing connections to become unauthorized at some point shortly after it is done; I don't know the specifics, as that's a service implementation detail.  By rolling the credential, you're essentially telling the service to disallow anything using the old key.

- When you provide a connection string to the client library, you're giving it a hard-coded key.  If you roll that key and need to update, you have to destroy your active Event Hub client instances and instantiate new ones with the new key.

- The issue that you linked was the result of some thought exercises on how we could support rolling a shared key in place without requiring the full-on "build another client" flow.  That said, it is also one of the big reasons that we're working hard to encourage developers to use an identity-based constructor for production applications.

- We do have a credential type that allows a shared key to be used as an identity, separate of a connection string.  We've got an [item](https://app.zenhub.com/workspaces/azure-sdk-team-5bdca72c4b5806bc2bf0aab2/issues/azure/azure-sdk-for-net/6618) to offer support on that to provide a new shared key on-the-fly, but developers would still need to do some work to plumb that into their code, where the AAD identity rolling is more transparent.

Did that help at all or just further muddle the issue?

albertodenatale replied:
Thank you very much for your answer that was amazing.

That task is very interesting.

albertodenatale replied:
> * but developers would still need to do some work to plumb that into their code, where the AAD identity rolling is more transparent

To see if I got it right, do you think this line would work?

> Another huge advantage compared to using Connection Strings is that rolling of credentials
> is a feature of OAuth 2.0 and implemented by Azure Active Directory. Whenever a token expires or is revoked, the SDK will take care of issuing a new one transparently. The same would not be true with connection strings and changing credentials will require the EventHubProducerClient or EventHubProducerConsumer to be disposed and recreated.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample13_AuthenticateWithClientSecretCredential.cs:
```diff
@@ -0,0 +1,71 @@
+
+            // ClientSecretCredential allows performing service principal authentication passing
+            // tenantId, clientId and clientSecret directly from the constructor.
+            // (see: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow)
```
Nicely done!    This is very clear and well-written.   Thank you for your great attention to detail here.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample13_AuthenticateWithClientSecretCredential.cs:
```diff
@@ -0,0 +1,71 @@
+
+            // EventHubClient takes ClientSecretCredential from its constructor and tries to issue a token from AAD.
+
+            await using (EventHubClient client = new EventHubClient(fullyQualifiedNamespace, eventHubName, credentials))
```
With the client type hierarchy overhaul, I'd recommend using the `EventHubProducerClient` here; the parameter set can be used without changes and the constructor signature is stable.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -15,6 +15,16 @@ namespace Azure.Messaging.EventHubs.Samples
     ///
     public static class Program
     {
+        private static string[] extraOptionsForSamples = new[]
```
nit:  If you would be so kind, for consistency, please add single line summary doc comments for these.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -15,6 +15,16 @@ namespace Azure.Messaging.EventHubs.Samples
+            "Explore Identity Samples"
+        };
+
+        private static string[] extraOptionsForIdentitySamples = new[]
```
I think we can make these `readonly` as well.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -137,6 +381,34 @@ private static void DisplayHelp()
+        ///
+        private static int? ReadSelection(IReadOnlyList<IEventHubsSample> samples)
+        {
+            return ReadSelection(samples.Count() + extraOptionsForSamples.Length);
```
nit:  Since this is a list, you can use the `Count` property directly.   Not a huge deal since the LINQ method will delegate to it, but I'd still recommend it.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -137,6 +381,34 @@ private static void DisplayHelp()
+        ///
+        /// <returns>The validated selection that was made.</returns>
+        ///
+        private static int? ReadSelection(IReadOnlyList<IEventHubsIdentitySample> samples)
```
You could make these into expression bodied methods, but I leave that to your discretion.   I don't have a strong opinion.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -15,6 +15,16 @@ namespace Azure.Messaging.EventHubs.Samples
     ///
     public static class Program
```
It may be helpful to add an example to the help text with the identity sample parameters set.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -251,6 +539,21 @@ private static CommandLineArguments ParseArguments(string[] args)
+        ///
+        /// <returns>The set of samples defined in the solution.</returns>
+        ///
+        private static IReadOnlyList<IEventHubsIdentitySample> LocateIdentitySamples() =>
```
You could make one of these generic and collapse to a single method.

albertodenatale replied:
Nice one :)

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -137,6 +381,34 @@ private static void DisplayHelp()
+        ///
+        /// <returns>The validated selection that was made.</returns>
+        ///
+        private static int? ReadSelection(IReadOnlyList<IEventHubsSample> samples)
```
You could collapse this down into one method if you made it generic.   Something like:

```csharp
private static int? ReadSelection<TSampleCategory>(IReadOnlyList<TSampleCategory> samples) => samples switch
{
    IReadOnlyList<IEventHubsSample> samples => ReadSelection(samples.Count + extraOptionsForSamples.Length),
    IReadOnlyList<IEventHubsIdentitySample> samples => return ReadSelection(samples.Count + extraOptionsForIdentitySamples.Length),
    _ => throw new InvalidArgumentException()
};
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -74,30 +229,119 @@ public static async Task Main(string[] args)
+        {
+            // Prompt for the Fully Qualified Namespace, if it wasn't passed.
+
+            while (String.IsNullOrEmpty(parsedArgs.FullyQualifiedNamespace))
```
nit:  The coding standards request that we use `string`, despite my preference for this style.    Looks like our analysis tool missed changing these during its last pass.  Do you mind doing a global replace to `string.IsNullOrEmpty` ?

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -42,27 +52,172 @@ public static async Task Main(string[] args)
+
+            for (int i = 0; i < extraOptionsForSamples.Length; i++)
+            {
+                Console.WriteLine($"{ samples.Count + i + 1 }) {extraOptionsForSamples[i]}");
```
nit:  If you would be so kind, would you put a space a between your `{` and the interpolated value?

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -42,27 +52,172 @@ public static async Task Main(string[] args)
+        ///
+        /// <returns>If the user has chosed to go back in the main sample listing.</returns>
+        ///
+        private static bool IsGoBack(IReadOnlyList<IEventHubsIdentitySample> identitySamples, int? choice)
```
You allow for a set of extra options, but then make a positional assumption based on the last option.    Should we either make this a text-based comparison or make the "extra options" a single member to avoid mistakes... or am I reading this wrong and missing a detail?

albertodenatale replied:
Yes, there is an important very "sneaky" detail here. My fault here if it was not evident my apologies for missing the required level of detail.

The indexing is zero-based and the right implementation would be:

```csharp

/// <summary>
///   It checks if an option is to go back.
/// </summary>
///
/// <param name="samples">A list of identity samples</param>
/// <param name="choice">The zero based index referring to the option chosen from console</param>
///
/// <returns>If the user has chosed to go back in the main sample listing.</returns>
///
private static bool IsGoBack(IReadOnlyList<IEventHubsIdentitySample> identitySamples, int? choice)
{
   return IsLastOption(identitySamples, choice);
}

private static bool IsLastOption(IReadOnlyList<IEventHubsIdentitySample> identitySamples, int? choice)
{
    return choice == (identitySamples.Count + extraOptionsForIdentitySamples.Count() - 1);
}

```

I needed to split the extra options into two different lists so to have to compute the maximum option allowed for each category.

albertodenatale replied:
I take you meant ```Length``` rather than ```Count```

jsquire replied:
Yup.  My bad.   I read that as the list coming in.   Sorry about that.

albertodenatale replied:
No worries mate, you did great

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -74,30 +229,119 @@ public static async Task Main(string[] args)
+        ///
+        /// <param name="parsedArgs">The arguments passed from console.</param>
+        ///
+        private static void PromptEventHubName(CommandLineArguments parsedArgs)
```
I love this approach!   One small nit, since it does a conditional check before prompting, it likely makes sense to revise to `PromptIfMissing` or similar?

jsquire reviewed (approved):
Thanks, Alberto!   This is very well thought out and solid.   I made a few suggestions, most of which were minor nits around formatting.

There are a couple of recent changes that may impact this when you rebase:

- Client types have been overhauled to create a more direct client-to-operation match.   This shouldn't be a bit adjustment and I made a suggestion that should be a small fix.

- The decision was made at a higher level to remove project-based references to `Core` and `Identity`, so you'll see the abstraction that we just collaborated on was removed in favor of a hard dependency.     Sorry about that,  I wasn't in the loop.

jsquire reviewed (commented):
Whoops.   I meant to submit as comment and accidentally approved.    I'd like to keep these in draft until we finish iterating and I can do a local test/run to validate.    Sorry about that.

jsquire reviewed (changes requested):
...

albertodenatale commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/TestEnvironment.cs:
```diff
@@ -113,6 +121,24 @@ public static class TestEnvironment
+
+        /// <summary>
+        ///   The name of the Azure Active Directory client secret of the service principal to use for testing
+        ///   AAD authentication and holding the 'Azure Event Hubs Data Owner' role.
```
Another AAD that will need removal

albertodenatale commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -42,27 +52,172 @@ public static async Task Main(string[] args)
+        ///
+        /// <returns>A sample or null if that could not be determined.</returns>
+        ///
+        private static INamedSample RetrieveSample()
```
> Since this contains only two members, I think it would be better to duplicate them in the interface for each sample category to keep things localized at a glance. Each category of sample will need its own means of running the samples that would have to have direct knowledge of its dependencies.

@jsquire 

That is a very good point. This interface helps to break the ```Main``` into two sections:
- Part 1, retrieve the sample
- Part 2, run the sample

We could remove the interface and have a struct like this:

```csharp

public struct Choice
{
      public int? Choice;
      public SampleCategory? SampleCategory;
}

public enum SampleCategory
{
     EventHubs,
     Identity
}

private static Choice RetrieveSample()
{
    ...
}

```

Essentially we would remove ```INamedSample``` and we would add an enum ```SampleCategory```?

Look forward to knowing your thoughts about it.

albertodenatale replied:
I love your idea, ISample is much more compact and simplifies a lot the naming of the interface.

> somehow it feels heavier than ideal either way.

What if, we introduced a Menu class like:

```csharp

using(var menu = new Menu())
{
   ISample sample = menu.RetrieveSample()

  // (... continues as it does today)
}

public class Menu: IDisposable
{
    private readonly IEnumerable<IEventHubsSample> Samples;
    private readonly IEnumerable<IEventHubsIdentitySample> IdentitySamples;

    public Menu()
    {
        Samples = LocateSamples<IEventHubsSample>();
        IdentitySamples = LocateSamples<IEventHubsIdentitySample>();
    }

    public ISample RetrieveSample()
    {
          // Here the logic we are using today
          // that prints the samples
    }
}

```

Another idea could be providing the user with arrow/color navigation? Not sure how easy that would be as I have never done it before.

jsquire replied:
I love the idea, but I'm hesitant to put too much more engineering effort into the runner.   Given some internal chatter about "how do we make samples the same across all of these libraries", it seems like there's going to be standardization on "package each sample as a console application" in the not too distant future.   Though it's not yet an actionable standard, it does give me pause about doing too much tuning over the sample runner user experience or design as we've got it now.   

I think our best course of action is to keep things as simple and close to what we've got, even if that's a bit less elegant than would be ideal for a crafted user experience.  I'm excited to get this sample into our nightly test run, as it will be the first coverage that we have on the true identity path.   I think that is a huge win, as it having the working sample itself for folks to see in action.

albertodenatale commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/TestEnvironment.cs:
```diff
@@ -36,6 +36,14 @@ public static class TestEnvironment
             new Lazy<string>(() => ReadAndVerifyEnvironmentVariable("EVENT_HUBS_SECRET"), LazyThreadSafetyMode.PublicationOnly);
 
+        /// <summary>The environment variable value for the Azure Active Directory client identifier of the service principal, lazily evaluated.</summary>
+        private static readonly Lazy<string> s_eventHubsIdentityClientInstance =
```
Great point. After working on the scripts in this [conversation](https://github.com/Azure/azure-sdk-for-net/pull/8545#discussion_r341874053), if we decided it is ok to have one principal with both the ```Contributor``` and the ```Azure Event Hubs Data Owner``` assigned to it, we may remove these variables completely?

albertodenatale commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Infrastructure/INamedSample.cs:
```diff
@@ -0,0 +1,25 @@
+    ///   Designates a class as a sample
+    /// </summary>
+    ///
+    public interface INamedSample
```
Definitely yes, added some points in [here](https://github.com/Azure/azure-sdk-for-net/pull/8470#discussion_r341871552)

jsquire reviewed (commented):
Looking solid!   I think we're pretty close; only a couple of minor things that seem to be pending.   Thanks for continuing to iterate!

albertodenatale commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -138,23 +429,39 @@ private static void DisplayHelp()
         /// <returns>The validated selection that was made.</returns>
         ///
-        private static int? ReadSelection(int sampleCount)
+        private static int? ReadSelection<TSampleCategory>(IReadOnlyList<TSampleCategory> samples) => samples switch
```
Indeed, they are very elegant, like C# is.

albertodenatale commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs:
```diff
@@ -42,27 +82,189 @@ public static async Task Main(string[] args)
-                parsedArgs.ConnectionString = Console.ReadLine().Trim();
+                Console.WriteLine();
+                Console.WriteLine();
+                Console.WriteLine("Quitting...");
```
The only real case when ```sample``` is null would not actually be if the sample when the user types 'X'. 

Do you find returning ```null``` on that instance misleading?

jsquire replied:
Nope, I think the `null` for "There was no sample chosen" scenario makes sense.   I was thinking more about "what happens if someone comes in an removes all of the samples, then tries to run this thing.    Granted, it's not a likely or expected thing, but differentiating between "We couldn't find anything" and "You've chosen to exit" was on my brain.

That said, given commentary elsewhere about how we're likely to be asked to restructure, I'd argue (with myself) that it doesn't make sense to code that up, given that it's an incredibly unlikely scenario and is more apt to indicate that someone made some pretty strange changes to the repo.

albertodenatale replied:
> what happens if someone comes in an removes all of the samples, then tries to run this thing.

That is a good point, I might do some refactoring and fix this relatively quickly?

jsquire replied:
I'd call it an extreme corner case; let's break that out into a separate PR if it's something that you'd like to do.  It's definitely less valuable than having your changes integrated.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/samples/Infrastructure/ISample.cs:
```diff
@@ -0,0 +1,25 @@
+    ///   Designates a class as a sample
+    /// </summary>
+    ///
+    public interface ISample
```
Yes!   That little tweak seems to make things a lot more clear when looking at the interface.   Nicely done!

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Samples/SamplesLiveTests.cs:
```diff
@@ -46,11 +59,33 @@ public class SamplesLiveTests
         public async Task SmokeTestASample(IEventHubsSample sample)
         {
-            await using (EventHubScope scope = await EventHubScope.CreateAsync(2))
+            await using (var scope = await EventHubScope.CreateAsync(2))
```
*slow clap*

Damn static analysis tool that was run really doesn't appreciate `var` and was overly aggressive about it.

jsquire reviewed (approved):
Things look solid.   Feel free to keep iterating if you've got things that you'd like to tweak, but I think we're at a good place to merge.

albertodenatale commented:
@jsquire 

Thank you, completed end to end tests and everything seems to run smoothly. Fixed a tiny bug I had introduced. Ready for merging when you are.

jsquire commented:
> @jsquire
> 
> Thank you, completed end to end tests and everything seems to run smoothly. Fixed a tiny bug I had introduced. Ready for merging when you are.
> 
> Do you think would be better to merge everything as a single commit to preserve the repo history?

I personally like shaping my commits to tell the story; not so much every single iteration but using a few if there's a pivot or something that would be of interest.   However.... our options are going to rebase and squish things unless settings have changed.

jsquire reviewed (approved):
> > @jsquire
> > Thank you, completed end to end tests and everything seems to run smoothly. Fixed a tiny bug I had introduced. Ready for merging when you are.
> > Do you think would be better to merge everything as a single commit to preserve the repo history?
> 
> I personally like shaping my commits to tell the story; not so much every single iteration but using a few if there's a pivot or something that would be of interest. However.... our options are going to rebase and squish things unless settings have changed.

Given this thread, I'll delay merging for you to decide how you'd like to represent your work, with the disclaimer that our system may take some liberties.   When you're ready, yell at me with a 👍

albertodenatale commented:
Thank you, did a squash commit and refreshed the test with the new interface 😃 ready whenever you are 👍
